### ISSUE TITLE
Could not analyze layer: Got response 400 with message

### ISSUE THREAD

**Issue by HackToday** [quay/clair#283]
1. Install clair with docker:
https://github.com/coreos/clair#docker

2. Install analyze tool  with
go get -u github.com/coreos/clair/contrib/analyze-local-images

3. Try analyze image:
```
# ./analyze-local-images  -endpoint http://host-ip:6060  -my-address hostip  09c911450d69
2016-12-07 08:54:25.528300 I | Saving 09c911450d69 to local disk (this may take some time)
2016-12-07 08:54:47.639785 I | Retrieving image history
2016-12-07 08:54:47.640177 I | Setting up HTTP server (allowing: 10.15.240.120)
2016-12-07 08:54:47.740349 I | Analyzing 14 layers... 
2016-12-07 08:54:47.740393 I | Analyzing d0fc2f8729503b3c51845d2f65ae8fb03c60b1bbebb7e42ea06890c5fd5a8012
2016-12-07 08:54:47.748294 I | Could not analyze layer: Got response 400 with message {"Error":{"Message":"could not find layer"}}

```

**Comment by HackToday** [comment]
@Quentin-M  Did you know what's wrong with it ?

**Comment by Quentin-M** [comment]
@HackToday This means that Clair couldn't get the layer tarball. I'd read the Clair logs to find out which URL was tried. That URL should contain the IP address of the machine that runs the analyze-local-images tool. Otherwise, the `my-address` is misused.

**Comment by HackToday** [comment]
Hi  @Quentin-M  As above comment, I installed clair with docker. 

I found that, with docker ways, it need to mount host tmp folder, which https://github.com/coreos/clair#docker not mentioned, but it was mentioned in https://github.com/coreos/clair/tree/master/contrib/analyze-local-images


Also I have one question, for the `endpoint` and `my-address` meaning (remote run clair way)

-endpoint "http://192.168.99.100:6060" -my-address "192.168.99.1"

If I installed clair on one machine, let's say machine A
and I run above commend in another machine, let's say machine B (machine B should have docker images, right ?)

So endpoint here is machine A ip,  my-address here is machine B ip


is that ? 

@Quentin-M  Could you help above questions ?

Thanks

**Comment by pandeybk** [comment]
I am seeing the same error when I tried to run clair in kubernetes cluster. Following findings so far,

I am running following command,

```
/home/go/gopath/bin/analyze-local-images -endpoint "<clair-server-loadbalancer-address>:6060" -my-address "<public-ip-with-9279-port-allowed-to-public" <docker-image>
```

And seeing following output in local machine 

```
2017-01-31 17:15:16.856078 I | Saving <docker-image> to local disk (this may take some time)
2017-01-31 17:15:55.714971 I | Retrieving image history
2017-01-31 17:15:55.715094 I | Setting up HTTP server (allowing: <loadbalancer-address>)
2017-01-31 17:15:55.815300 I | Analyzing 13 layers... 
2017-01-31 17:15:55.815320 I | Analyzing 7f1d5a7e7c722d89fb14528ffde7ef6002c737cd4c5e694306e67934f8baee3e
2017-01-31 17:15:56.187778 I | Could not analyze layer: Got response 400 with message {"Error":{"Message":"could not find layer"}}
```

However layer was created successfully in local machine 
```
[root@ip-10-0-0-197 analyze-local-image-086374639]# ls -la
total 36
drwx------ 9 root root 4096 Jan 31 16:58 .
drwxr-xr-x 3 root root 4096 Jan 31 16:58 ..
drwxr-xr-x 2 root root 4096 Dec 23 16:46 067a767eaa01e63621eda6846184ec9f96494d5bf06a7a11b9bda8a4ad11b9e2
drwxr-xr-x 2 root root 4096 Dec 23 16:46 1955e548b94199cfb33c4c1981589ea3f6d292e451f0686e3d102c029493ff34
drwxr-xr-x 2 root root 4096 Dec 23 16:46 3ad84667cd29905eda0607e2376462d07d7fa7543e2f9ad8db92ae32c6702f4d
drwxr-xr-x 2 root root 4096 Dec 23 16:46 5cacc871d7c0290e7fccdf3899af969ae9a271eb926e9a2290bbf5379b6d9da9
drwxr-xr-x 2 root root 4096 Dec 23 16:46 6f1453466f47b3fd280d71f71f2447219e38ab0e116b2f685d1f9fbe83069803
drwxr-xr-x 2 root root 4096 Dec 23 16:46 79631551da9cb10b53a7149aeff805f65ca315d820204fbfb9b68a05b0224370
drwx------ 2 root root 4096 Jan 31 16:58 7f1d5a7e7c722d89fb14528ffde7ef6002c737cd4c5e694306e67934f8baee3e
```

If you look at the pods logs, I am seeing following:

```
2017-01-31 17:15:56.133936 W | detectors: could not download layer: got status code 403, expected 2XX
2017-01-31 17:15:56.133980 E | worker: layer 7f1d5a7e7c722d89fb14528ffde7ef6002c737cd4c5e694306e67934f8baee3e: failed to extract data from http://<public-ip>:9279/7f1d5a7e7c722d89fb14528ffde7ef6002c737cd4c5e694306e67934f8baee3e/layer.tar: could not find layer
2017-01-31 17:15:56.134433 I | api: 10.9.94.147:1494 "POST /v1/layers" 400 (183.489698ms)
```

**Comment by Quentin-M** [comment]
@pandeybk This tool is all but meant to do anything else than scanning local images on your own machine, against a single Clair Instance, to discover what insights can Clair provide. The logs shows that the local web server, started by the tool to serve the image to Clair, only allows the load-balancer's public IP to access the layer. However Clair, behind the load-balancer accesses the server using its own IP, which is not allowed.

**Comment by pandeybk** [comment]
make sense @Quentin-M thank you for reply.

**Comment by jmellor** [comment]
I may have the same issue, being unable to pull a layer to analyze causing Clair to fail.  About half of my builds fail to analyse for this reason.  However in my case, the cause appears to be that the missing layer was compressed out upstream.  All of my failures have a base FROM image being the stock upstream Ubuntu images, so I have no control over how they are mis-built.

For an easy-to-reproduce example, if I "docker pull ubuntu:xenial-latest" and run it through Clair, it aborts.

Can we get a high priority fix in to ignore missing layers and analyze anyway?

**Comment by jmellor** [comment]
My previous attempts have been to use Clair to analyse an image from the registry, so I altered one Jenkins job to analyse a local image.  This also fails in the same way, even though the image works correctly.  The problem appears to be in Clair.  Critical fault.

Typical local analysis failure:
00:10:58.599 + docker push registry.internal/soctools/cookie-bro
00:10:58.612 The push refers to a repository [registry.internal/soctools/cookie-bro]
00:10:58.624 78b046751c18: Preparing
00:10:58.624 e8dac7387fbe: Preparing
00:10:58.624 0a05f0c11797: Preparing
00:10:58.624 417a4203d421: Preparing
00:10:58.624 ea2821e17568: Preparing
00:10:58.624 79717468825f: Preparing
00:10:58.624 6d1cb131e81a: Preparing
00:10:58.624 374eda78f988: Preparing
00:10:58.624 cf6d1cd028ea: Preparing
00:10:58.624 3d1126efbbe3: Preparing
00:10:58.624 6d1cb131e81a: Waiting
00:10:58.624 374eda78f988: Waiting
00:10:58.624 cf6d1cd028ea: Waiting
00:10:58.624 3d1126efbbe3: Waiting
00:10:58.624 79717468825f: Waiting
00:10:58.780 417a4203d421: Pushed
00:10:58.844 79717468825f: Layer already exists
00:10:58.931 6d1cb131e81a: Layer already exists
00:10:58.999 374eda78f988: Layer already exists
00:10:59.061 cf6d1cd028ea: Layer already exists
00:10:59.145 3d1126efbbe3: Layer already exists
00:11:00.499 e8dac7387fbe: Pushed
00:11:00.939 0a05f0c11797: Pushed
00:11:06.453 78b046751c18: Pushed
00:11:21.833 ea2821e17568: Pushed
00:11:21.954 latest: digest: sha256:a02734aeb9e9e3267e41a82af4e67fa43a5890113d97c907aafa3aee1c94cbeb size: 2414
00:11:21.956 + /var/lib/jenkins/bin/analyze-local-images -endpoint http://10.203.0.16:6060 -my-address 10.203.0.36 registry.internal/soctools/cookie-bro
00:11:21.965 2017-02-02 12:09:51.095322 I | Saving registry.internal/soctools/cookie-bro to local disk (this may take some time)
00:11:30.850 2017-02-02 12:09:59.983583 I | Retrieving image history
00:11:30.850 2017-02-02 12:09:59.983774 I | Setting up HTTP server (allowing: 10.203.0.16)
00:11:30.950 2017-02-02 12:10:00.084022 I | Analyzing 10 layers... 
00:11:30.950 2017-02-02 12:10:00.084062 I | Analyzing 54375dd70b96afbbc47f2c95dceecfe4c937f45d73f1795d6ef52dc52872a1ed
00:11:30.957 2017-02-02 12:10:00.091007 I | Could not analyze layer: Got response 400 with message {"Error":{"Message":"could not find layer"}}

Huh?  Clair is doing the wrong thing!
